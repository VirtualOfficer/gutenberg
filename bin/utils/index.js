/**
 * External dependencies
 */
const fs = require( 'fs' );
const childProcess = require( 'child_process' );
const inquirer = require( 'inquirer' );
const chalk = require( 'chalk' );

const error = chalk.bold.red;

/**
 * Small utility used to read an uncached version of a JSON file
 *
 * @param {string} fileName
 */
function readJSONFile( fileName ) {
	const data = fs.readFileSync( fileName, 'utf8' );
	return JSON.parse( data );
}

/**
 * Utility to run a child script
 *
 * @param {string} script Script to run.
 * @param {string} cwd    Working directory.
 */
function runShellScript( script, cwd ) {
	childProcess.execSync( script, {
		cwd,
		env: {
			NO_CHECKS: true,
			PATH: process.env.PATH,
			HOME: process.env.HOME,
		},
		stdio: [ 'inherit', 'ignore', 'inherit' ],
	} );
}

/**
 * Asks the user for a confirmation to continue or abort otherwise
 *
 * @param {string} message      Confirmation message.
 * @param {boolean} isDefault   Default reply.
 * @param {string} abortMessage Abort message.
 */
async function askForConfirmationToContinue( message, isDefault = true, abortMessage = 'Aborting.' ) {
	const { isReady } = await inquirer.prompt( [ {
		type: 'confirm',
		name: 'isReady',
		default: isDefault,
		message,
	} ] );

	if ( ! isReady ) {
		/* eslint-disable no-console */
		console.log( error( '\n' + abortMessage ) );
		/* eslint-enable no-console */
		process.exit( 1 );
	}
}

module.exports = {
	readJSONFile,
	runShellScript,
	askForConfirmationToContinue,
};
