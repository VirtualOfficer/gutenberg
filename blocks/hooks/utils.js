/**
 * WordPress dependencies
 */
import { isValidElement, Children, cloneElement } from '@wordpress/element';

export function mapElements( children, cb ) {
	return Children.map( children, child => {
		if ( ! isValidElement( child ) ) {
			return child;
		}

		const newChild = cb( child );
		// If a new child is returned then we should not traverse its children
		if ( newChild === child ) {
			const childProps = { ...child.props, children: mapElements( child.props.children, cb ) };
			return cloneElement( child, childProps );
		}

		return newChild;
	} );
}
