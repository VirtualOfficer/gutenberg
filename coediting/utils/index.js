/**
 * External dependencies
 */
import { camelCase, isObject } from 'lodash';

export function camelCaseKeysDeep( input ) {
	if ( Array.isArray( input ) ) {
		return input.map( camelCaseKeysDeep );
	}

	if ( isObject( input ) ) {
		return Object.keys( input ).reduce( ( acc, key ) => {
			acc[ camelCase( key ) ] = camelCaseKeysDeep( input[ key ] );
			return acc;
		}, {} );
	}

	return input;
}
