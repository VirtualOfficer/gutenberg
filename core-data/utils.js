/**
 * External dependencies
 */
import { isPlainObject } from 'lodash';

/**
 * Module constants
 */
const DEFAULT_ID = '[[default]]';

/**
 * Generates a unique identifier based on an object/scalar query arg.
 *
 * @param  {*}      query
 *
 * @return {string}       Request ID.
 */
export function getRequestId( query = DEFAULT_ID ) {
	return isPlainObject( query ) ? JSON.stringify( query ) : query;
}
