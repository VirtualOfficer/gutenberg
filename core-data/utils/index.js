// Temporary: `queried-data` is modeled as an indepent module, so while its
// implementation is reused here, the source location may change.
export { default as onSubKey } from '../queried-data/utils/on-sub-key';
