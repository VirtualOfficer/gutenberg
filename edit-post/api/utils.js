/* eslint no-console: [ 'error', { allow: [ 'error' ] } ] */

/**
 * Validates the plugin id.
 *
 * @param {string} id Plugin identifier.
 * @return {boolean} Whether the pluginId is valid.
 */
export function validatePluginId( id ) {
	if ( typeof id !== 'string' ) {
		console.error(
			'Plugin identifier must be a string.'
		);
		return false;
	}
	if ( ! /^[a-z][a-z0-9-]*\/[a-z][a-z0-9-]*$/.test( id ) ) {
		console.error(
			'Plugin identifier must contain a namespace prefix, include only lowercase alphanumeric characters or dashes, and start with a letter. Example: my-plugin/my-custom-sidebar'
		);
		return false;
	}
	return true;
}
