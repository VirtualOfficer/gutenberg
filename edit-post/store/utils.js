/**
 * WordPress dependencies
 */
import { subscribe } from '@wordpress/data';

/**
 * Given a selector returns a functions that returns the listener only
 * if the returned value from the selector changes.
 *
 * @param  {function} selector Selector.
 * @param  {function} listener Listener.
 * @return {function}          Listener creator.
 */
export const onChangeListener = ( selector, listener ) => {
	let previousValue = selector();
	return () => {
		const selectedValue = selector();
		if ( selectedValue !== previousValue ) {
			previousValue = selectedValue;
			listener( selectedValue );
		}
	};
};

/**
 * Given a selector, the desired value and a listener,
 * the function calls the listener when the selector matches the passed value.
 * If the selector does not matches the value right away the function subscribes to data module changes.
 *
 * @param  {function} selector Selector.
 * @param  {*} value           The value to compare against the selector result.
 * @param  {function} listener Listener to be called when selector matches value.
 */
export const onValueMatch = ( selector, value, listener ) => {
	if ( selector() === value ) {
		listener();
		return;
	}
	const unsubscribe = subscribe( () => {
		if ( selector() === value ) {
			unsubscribe();
			listener();
		}
	} );
};
