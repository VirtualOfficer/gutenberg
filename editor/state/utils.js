/**
 * Returns a post attribute value, flattening nested rendered content using its
 * raw value in place of its original object form.
 *
 * @param  {*} value Original value
 * @return {*}       Raw value
 */
export function getPostRawValue( value ) {
	if ( value && 'object' === typeof value && 'raw' in value ) {
		return value.raw;
	}

	return value;
}
