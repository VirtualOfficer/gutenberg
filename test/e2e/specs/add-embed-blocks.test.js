/**
 * Internal dependencies
 */
import '../support/bootstrap';
import { newPost, newDesktopBrowserPage } from '../support/utils';

describe( 'embedding media', () => {
	beforeEach( async () => {
		await newDesktopBrowserPage();
		await newPost();
		await page.click( '.edit-post-header [aria-label="Add block"]' );
		await page.keyboard.type( 'embed' );
		await page.keyboard.press( 'Tab' );
		await page.keyboard.press( 'Tab' );
		await page.keyboard.press( 'Enter' );
	} );

	afterEach( async () => {
		//reset view to visual editor for next test
		await page.click( '.edit-post-more-menu [aria-label="More"]' );
		const visualEditorButton = ( await page.$x( '//button[contains(text(), \'Visual Editor\')]' ) )[ 0 ];
		await visualEditorButton.click( 'button' );
	} );

	it( 'Should insert an embedded tweet', async () => {
		//insert a twitter link
		await page.keyboard.type( 'https://twitter.com/WordPress/status/980716623060455424' );
		await page.keyboard.press( 'Tab' );
		await page.keyboard.press( 'Enter' );

		//wait for link to resolve and block to rerender
		await page.waitFor( 2000 );

		//Check HTML output
		await page.click( '.edit-post-more-menu [aria-label="More"]' );
		const codeEditorButton = ( await page.$x( '//button[contains(text(), \'Code Editor\')]' ) )[ 0 ];
		await codeEditorButton.click( 'button' );

		//Assert that the editor now has an embedded tweet block
		const textEditorContent = await page.$eval( '.editor-post-text-editor', ( element ) => element.value );
		expect( textEditorContent ).toMatchSnapshot();
	} );

	it( 'Should insert a Youtube video', async () => {
		//insert a youtube link
		await page.keyboard.type( 'https://www.youtube.com/watch?v=CiwJx557ttk' );
		await page.keyboard.press( 'Tab' );
		await page.keyboard.press( 'Enter' );

		//wait for link to resolve and block to rerender
		await page.waitFor( 2000 );

		//Check HTML output
		await page.click( '.edit-post-more-menu [aria-label="More"]' );
		const codeEditorButton = ( await page.$x( '//button[contains(text(), \'Code Editor\')]' ) )[ 0 ];
		await codeEditorButton.click( 'button' );

		//Assert that the editor now has an embedded youtube block
		const textEditorContent = await page.$eval( '.editor-post-text-editor', ( element ) => element.value );
		expect( textEditorContent ).toMatchSnapshot();
	} );
} );
