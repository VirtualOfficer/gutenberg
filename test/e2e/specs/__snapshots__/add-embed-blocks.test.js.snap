// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`embedding media Should insert a Youtube video 1`] = `
"<!-- wp:embed {\\"url\\":\\"https://www.youtube.com/watch?v=CiwJx557ttk\\",\\"type\\":\\"video\\",\\"providerNameSlug\\":\\"youtube\\"} -->
<figure class=\\"wp-block-embed is-type-video is-provider-youtube\\">
    https://www.youtube.com/watch?v=CiwJx557ttk
</figure>
<!-- /wp:embed -->"
`;

exports[`embedding media Should insert an embedded tweet 1`] = `
"<!-- wp:embed {\\"url\\":\\"https://twitter.com/WordPress/status/980716623060455424\\",\\"type\\":\\"rich\\",\\"providerNameSlug\\":\\"twitter\\"} -->
<figure class=\\"wp-block-embed is-type-rich is-provider-twitter\\">
    https://twitter.com/WordPress/status/980716623060455424
</figure>
<!-- /wp:embed -->"
`;
